OpenKh.Bbs/Pmo.cs

```diff
@@ -0,0 +1,716 @@
+                            stream.Seek(vertexIncreaseAmount, SeekOrigin.Current);
+
+                            currentVertex.X = stream.ReadFloat();
+                            currentVertex.Y = stream.ReadFloat();
+                            currentVertex.Z = stream.ReadFloat();
+                            meshChunk.vertices.Add(currentVertex);
+                            break;
+                    }
+
+                    stream.Seek(vertexStartPos + meshChunk.SectionInfo.VertexSize, SeekOrigin.Begin);
+
+                    if (flags.Primitive == PrimitiveType.PRIMITIVE_TRIANGLE)
+                    {
+                        meshChunk.Indices.Add(v);
+                    }
+                }
+
+                VertCnt = meshChunk.SectionInfo.VertexCount;
+                pmo.Meshes.Add(meshChunk);
+
+                stream.Seek(positionAfterHeader + (meshChunk.SectionInfo.VertexCount * meshChunk.SectionInfo.VertexSize), SeekOrigin.Begin);
+                stream.Seek(stream.Position % 4, SeekOrigin.Current);
+            }
+        }
+
+        public static Pmo Read(Stream stream)
+        {
+            Pmo pmo = new Pmo();
+
+            ReadHeader(ref stream, ref pmo);
+            ReadTextureSection(ref stream, ref pmo);
+
+            // Read all data
+            UInt16 VertCnt = 0xFFFF;
+            int meshSecCount = 0;
+
+            ReadMeshData(ref stream, ref pmo, ref VertCnt, ref meshSecCount, 0);
+
+            stream.Seek(pmo.header.MeshOffset1, SeekOrigin.Begin);
+            VertCnt = 0xFFFF;
+            
+            if(pmo.header.MeshOffset1 != 0)
+            {
+                ReadMeshData(ref stream, ref pmo, ref VertCnt, ref meshSecCount, 1);
+            }
+
+            // Read textures.
+            for (int i = 0; i < pmo.textureInfo.Length; i++)
+            {
+                stream.Seek(pmo.textureInfo[i].TextureOffset + 0x10, SeekOrigin.Begin);
+                uint tm2size = stream.ReadUInt32() + 0x10;
+                stream.Seek(pmo.textureInfo[i].TextureOffset, SeekOrigin.Begin);
+
+                byte[] TextureBuffer = new byte[tm2size];
+                TextureBuffer = stream.ReadBytes((int)tm2size);
+
+                pmo.texturesData.Add(TextureBuffer);
+            }
+
+            // Read Skeleton.
+            if(pmo.header.SkeletonOffset != 0)
+            {
+                stream.Seek(pmo.header.SkeletonOffset, SeekOrigin.Begin);
+                pmo.skeletonHeader = BinaryMapping.ReadObject<SkeletonHeader>(stream);
+                pmo.jointList = new JointData[pmo.skeletonHeader.JointCount];
+                for (int j = 0; j < pmo.skeletonHeader.JointCount; j++)
+                {
+                    pmo.jointList[j] = BinaryMapping.ReadObject<JointData>(stream);
+                }
+            }
+
+            return pmo;
+        }
+
+        public static void Write(Stream stream, Pmo pmo)
+        {
+            BinaryWriter wr = new BinaryWriter(stream);
```

Comment:
I noticed you are using `BinaryWriter` to write some `sbyte` and `float` values. There are some helpers that do that for you in `OpenKh.Common` and there is not really need to use `BinaryWriter`. Just call `stream.Write(yourFloat)` and it should do it for you.

Reply:
```stream.Write(yourFloat)``` doesn't seem to work, which is why I went for BinaryWriter. I have to do ```OpenKh.Common.StreamExtensions.Write(yourFloat)``` for it to work. Or without ```OpenKh.Common``` in my case, since I'm using ```using OpenKh.Common```.